File: kotlin/examples/src/main/kotlin/Complex.kt
package gr.elaevents.buffela.examples

import gr.elaevents.buffela.deserialization.DeserializerBuffer
import gr.elaevents.buffela.examples.authToken.AuthTokenPayload
import gr.elaevents.buffela.examples.authToken.AuthTokenSignature
import gr.elaevents.buffela.examples.authToken.Date
import gr.elaevents.buffela.examples.authToken.Gender
import gr.elaevents.buffela.examples.authToken.User
import gr.elaevents.buffela.serialization.SerializerBuffer

fun main() {
    // Serialization ===================================================================================================

    val serializerBuffer = SerializerBuffer()

    val payload = AuthTokenPayload(
        issuedAt = System.currentTimeMillis().toDouble(),
        user = User.Registered.Viewer(
            userId = "588809b0-d8ce-4a6b-a2aa-9b10fd9d7a11",
            verified = true,
            birthDate = Date(
                year = 2003,
                month = 7,
                day = 22
            ),
            countryCode = 30u,
            phone = "1234567890",
            gender = Gender.MALE
        )
    )

    payload.serialize(serializerBuffer)

    serializerBuffer.clearBitBuffer()

    val signature = AuthTokenSignature(hmac256 = sign(serializerBuffer.toBytes()))
    signature.serialize(serializerBuffer)

    val serialized = serializerBuffer.toBytes()
    println("Serialized ${serialized.size} bytes")
    println(prettyByteArray(serialized))

    // Deserialization =================================================================================================

    val deserializerBuffer = DeserializerBuffer(serialized)

    val deserializedPayload = AuthTokenPayload.deserialize(deserializerBuffer)
    println("Deserialized payload: $deserializedPayload")

    deserializerBuffer.clearBitBuffer()

    val serializedPayload = serialized.sliceArray(0 until deserializerBuffer.position)
    val deserializedSignature = AuthTokenSignature.deserialize(deserializerBuffer)
    assertSigned(serializedPayload, deserializedSignature.hmac256)

    println("Signature is valid")
}

File: js/examples/buffela/complex.js
const { parseSchema } = require('@buffela/parser')
const { registerSerializer, SerializerBuffer } = require('@buffela/serializer')
const { registerDeserializer, DeserializerBuffer } = require('@buffela/deserializer');

const { sign, assertSigned } = require("../utils/signatureUtils");
const { prettyBuffer, prettyObject } = require("../utils/formatUtils");

const schemaJson = require('./AuthToken.json')

// Setup ===============================================================================================================

/** @type {import('./AuthToken').default} */
const schema = parseSchema(schemaJson)

registerSerializer(schema, {
    Date: {
        serialize(buffer, value) {
            const yearMonth = value.year * 12 + (value.month - 1)
            const day = value.day - 1

            buffer.writeUnsigned(yearMonth, 17)
            buffer.writeUnsigned(day, 5)
        }
    }
})

registerDeserializer(schema, {
    Date: {
        deserialize(buffer) {
            const yearMonth = buffer.readUnsigned(17)
            const day = buffer.readUnsigned(5)

            return {
                year: Math.floor(yearMonth / 12),
                month: yearMonth % 12 + 1,
                day: day + 1
            }
        }
    }
})


// Serialization =======================================================================================================

const serializerBuffer = new SerializerBuffer()

schema.AuthTokenPayload.serialize({
    issuedAt: Date.now(),
    user: {
        userId: '588809b0-d8ce-4a6b-a2aa-9b10fd9d7a11',
        User_type: schema.User.Registered,
        verified: true,
        Registered_type: schema.User.Registered.Viewer,
        birthDate: { year: 2003, month: 7, day: 22 },
        countryCode: 30,
        phone: '1234567890',
        gender: schema.Gender.MALE
    }
}, serializerBuffer)

serializerBuffer.clearBitBuffer()

schema.AuthTokenSignature.serialize({
    hmac256: sign(serializerBuffer.toBytes())
}, serializerBuffer)

const serialized = serializerBuffer.toBytes()

console.log('Serialized', serialized.length, 'bytes')
console.log(prettyBuffer(serialized))


// Deserialization =====================================================================================================

const deserializerBuffer = new DeserializerBuffer(serialized)

const payload = schema.AuthTokenPayload.deserialize(deserializerBuffer)
console.log('Auth token payload:', prettyObject(payload))

deserializerBuffer.clearBitBuffer()

const payloadBytes = serialized.subarray(0, deserializerBuffer.position)
const signature = schema.AuthTokenSignature.deserialize(deserializerBuffer)
assertSigned(payloadBytes, signature.hmac256)

console.log('Signature is valid')
